refactor(view): introduce AppViewManager to consolidate logic

tbosch opened the pull request:
AppViewManager is the single entry point to changing the view hierarchy.
It is split between the manager itself which does coordination and
helper methods, so both are easily testable in isolation.

Part of #1477 

After this is done the same pattern should be applied to the render side as well!

tbosch committed:
fix(test_lib): support multi matches with deep equality for function calls

rkirov commented on modules/angular2/src/core/compiler/dynamic_component_loader.js:
```diff
@@ -1,12 +1,11 @@
-import {ViewFactory} from 'angular2/src/core/compiler/view_factory';
-import {AppViewHydrator} from 'angular2/src/core/compiler/view_hydrator';
-import {ElementRef, DirectiveBinding} from './element_injector';
+import {AppViewManager} from 'angular2/src/core/compiler/view_manager';
```
General comment (no action required), our use of 'App' is in consistent, here it is meant as an opposite to 'render', but here: 
https://github.com/angular/angular/blob/ada1e642c57f19ad5cecde57db1aca1727bf2325/modules/angular2/src/core/application.js#L58
it is meant as a 'top-level'.

rkirov commented on modules/angular2/src/core/compiler/view_container.js:
```diff
@@ -6,103 +6,68 @@ import {isPresent, isBlank} from 'angular2/src/facade/lang';
+              defaultProtoView: viewModule.AppProtoView) {
+    this._viewManager = viewManager;
+    this._parentView = parentView;
+    this._boundElementIndex = boundElementIndex;
```
It is odd that `ViewContainer` doesn't have a ref to `InternalViewContainer` but instead has to go through an index and a list in parentView. Any reason for this decision?

tbosch replied:
We create the `InternalViewContainer` lazily when the first `AppView` is inserted into it. Before this, we don't need it...

rkirov commented on modules/angular2/src/core/compiler/view_manager.js:
```diff
@@ -0,0 +1,177 @@
+    this._utils = utils;
+  }
+
+  createDynamicComponentView(hostView:viewModule.AppView, boundElementIndex:number,
```
Why not use `ElementRef` instead of `boundElementIndex`?

tbosch replied:
`ElementRef` contains a lot more than we need here...

rkirov commented on modules/angular2/src/core/compiler/view_manager.js:
```diff
@@ -0,0 +1,177 @@
+      for (var binderIdx = 0; binderIdx < binders.length; binderIdx++) {
+        var binder = binders[binderIdx];
+        if (binder.hasStaticComponent()) {
+          var childView = this._createView(binder.nestedProtoView);
```
Shouldn't this be '_createViewRecurse' since it recurses here?

tbosch replied:
Good point.

rkirov commented on modules/angular2/src/core/compiler/view_manager_utils.js:
```diff
@@ -0,0 +1,243 @@
+    parentView.changeDetector.addChild(view.changeDetector);
+    var viewContainer = parentView.viewContainers[boundElementIndex];
+    if (isBlank(viewContainer)) {
+      viewContainer = new viewModule.InternalAppViewContainer();
```
Why are we building it on demand. Shouldn't all containers be known in advance as injectable parameters to decorators, etc?

tbosch replied:
They can be asked for via `ElementRef.viewContainer` for an element that does not inject a `ViewContainer` itself (needed for overlay service).

rkirov commented on modules/angular2/src/core/compiler/view_manager_utils.js:
```diff
@@ -0,0 +1,243 @@
+    if (isBlank(appInjector)) {
+      appInjector = hostElementInjector.getShadowDomAppInjector();
+    }
+    if (isBlank(appInjector)) {
```
why is this line necessary? Can we make `hostElementInjector.getShadowDomAppInjector()` always return non-null?

tbosch replied:
This is for letting a call overwrite the default injector (i.e. prevent the use of `hostElementInjector.getShadowDomAppInjector()`). Needed e.g. for creating a `ProtoView` in a `ViewContainer` with a pre filled injector (e.g. the router does this).

rkirov commented on modules/angular2/src/core/compiler/view_manager_utils.js:
```diff
@@ -0,0 +1,243 @@
+
+  _setUpEventEmitters(view:viewModule.AppView, elementInjector:eli.ElementInjector, boundElementIndex:number) {
+    var emitters = elementInjector.getEventEmitterAccessors();
+    for(var directiveIndex = 0; directiveIndex < emitters.length; ++directiveIndex) {
```
`for (var`

rkirov commented on modules/angular2/src/test_lib/test_bed.js:
```diff
@@ -112,12 +103,12 @@ export class TestBed {
  * Proxy to `AppView` return by `createView` in {@link TestBed} which offers a high level API for tests.
  */
 export class ViewProxy {
```
Not in this PR, but can we replace this test specific `ViewProxy` with a general `AppViewRef`. The apis used for tests should be similar to what we provide to users.

rkirov commented on modules/angular2/src/test_lib/test_lib.dart:
```diff
@@ -87,7 +91,32 @@ class Expect extends gns.Expect {
+      _expect(_firstArgsMatch(actual, a, b, c, d, e, f), true, reason: 'method invoked with correct arguments');
   Function get _expect => gns.guinness.matchers.expect;
+
+  // TODO(vsavkin): move this hack into Guinness
```
@vsavkin can you take a look at this.

rkirov commented on modules/angular2/test/core/compiler/dynamic_component_loader_spec.js:
```diff
@@ -24,7 +24,7 @@ import {If} from 'angular2/src/directives/if';
 
 export function main() {
-  describe('DynamicComponentLoader', function () {
+  ddescribe('DynamicComponentLoader', function () {
```
remove `ddescribe`

rkirov commented on modules/angular2/test/core/compiler/integration_spec.js:
```diff
@@ -38,7 +38,7 @@ import {ElementRef} from 'angular2/src/core/compiler/element_injector';
 
 export function main() {
-  describe('integration tests', function() {
+  ddescribe('integration tests', function() {
```
same

rkirov commented on modules/angular2/test/core/compiler/view_manager_spec.js:
```diff
@@ -0,0 +1,476 @@
+        binders = [];
+      }
+      var staticChildComponentCount = 0;
+      for (var i=0; i<binders.length; i++) {
```
syntax for iteration should be `for (var i = 0; i < binders.length; i++)`

tbosch committed:
refactor(view): introduce AppViewManager to consolidate logic

AppViewManager is the single entry point to changing the view hierarchy.
It is split between the manager itself which does coordination and
helper methods, so both are easily testable in isolation.

Also, ViewContainer is now only a pure reference to a bound element
with the previous functionality but does not contain the list of views
any more.

Part of #1477
